Use -fPIC by default

wilzbach opened the pull request:
The [release dmd.conf](https://github.com/dlang/dmd/blob/master/ini/linux/bin64/dmd.conf#L5) already contains `-fPIC`, but the default generated one doesn't.
This requires `libphobos2.a` to be built with `-fPIC`, but that one is already in the queue: https://github.com/dlang/phobos/pull/5586

Other PRs:

- [x] https://github.com/dlang/druntime/pull/1880
- [x] https://github.com/dlang/phobos/pull/5586

wilzbach committed:
Use -fPIC by default

wilzbach commented:
Seems like the header file generation on some platforms depends on `-fPIC`?

```
 ... compilable/testheader2i.d      -o- -H -Hfgenerated/compilable/header2i.di -inline ()
Test failed.  The logged output:
Executing post-test script: compilable/extra-files/header-postscript.sh header2i generated/ee125edca9bbbd1a4e2bb7d229beebf5e45342dfd0e3d8a6ad192a759807b8e2
120c120
< 	if (const n = 1)
---
> 	if (const @trusted n = 1)
129c129
< 	if (const shared n = 1)
---
> 	if (const shared @trusted n = 1)
../src/dmd -conf= -m32 -Icompilable -o- -H -Hfgenerated/compilable/header2i.di -inline  -odgenerated/compilable -ofgenerated/compilable/testheader2i_0.o -c compilable/testheader2i.d compilable/extra-files/header2.d


==============================
Test failed: expected rc == 0, exited with rc == 1

gmake[2]: *** [generated/compilable/testheader2i.d.out] Error 1
gmake[2]: *** Waiting for unfinished jobs....
 ... compilable/test12624.d         -lib -Icompilable/extra-files ()
gmake[1]: *** [start_compilable_tests] Error 2
gmake[1]: *** Waiting for unfinished jobs....
```

MartinNowak commented on src/posix.mak:
```diff
@@ -417,10 +417,10 @@ endif
 define DEFAULT_DMD_CONF
 [Environment32]
-DFLAGS=-I%@P%/../../../../../druntime/import -I%@P%/../../../../../phobos -L-L%@P%/../../../../../phobos/generated/$(OS)/release/32$(if $(filter $(OS),osx),, -L--export-dynamic)
+DFLAGS=-I%@P%/../../../../../druntime/import -I%@P%/../../../../../phobos -L-L%@P%/../../../../../phobos/generated/$(OS)/release/32$(if $(filter $(OS),osx),, -L--export-dynamic) -fPIC
```
Not for i386, it's too expensive there.

MartinNowak committed:
compile dmd with PIC on x86_64

- useful for compiler-as-library
- fixes compilation on default-PIE platforms
- use MODEL=32/64 to decide on default
- allow to override default with explicit PIC=0/1

MartinNowak committed:
small fixup
